### ISSUE TITLE
Investigate: Find all calls into the Flux library

### ISSUE THREAD

**Issue by jsternberg** [influxdata/influxdb#17789]
As part of our versioning experiment, we are working on an API for versioned Flux. The current idea is that each major version of Flux would have its own entrypoint and its own function signatures. To determine what influxdb requires, we want to prototype a single package that implements all of the external calls into Flux to help determine what is needed externally.

This should be modeled off of [this code](https://github.com/influxdata/influxdb/blob/8d2ba699a09c6064853341f18f8b7123e0aea6c7/query.go#L23) and that interface may be a good starting point.

This only covers *calls* into Flux code and does not cover usage of objects. For example, there may be some code that accepts an `*ast.Package` and modifies it or returns something from it. That code is not covered under this PR since it only uses the `*ast.Package`.

This issue covers invocation of free methods (ones that are not attached to a struct or interface) and the creation of structs.

```go
// This code would not be covered since it does not use any free functions and it does not create
// any new objects.
func NodeType(n ast.Node) string {
    return n.Type()
}

func RunMyScript(text string) {
    // This is an invocation of a free method and is covered by this issue.
    pkg := flux.Parse(text)
    runScript(pkg)
}

func MakeCompiler() flux.Compiler {
    // This is covered because it constructs a struct.
    return &lang.FluxCompiler{}
}
```

The output of this issue should be a package that encapsulates all of the calls into Flux so we can easily track how Flux is invoked. We may or may not merge this, but the existence of a PR demonstrating what it looks like is very important for this investigation.

**Comment by adrian-thurston** [comment]
Tried using `callgraph`  from `go tools` to see what is being called. After building callgraph:

`../tools/callgraph ./cmd/influxd > callgraph`

It shows a bunch of triples like:

`(github.com/influxdata/influxdb/v2/http.QueryRequest).analyzeFluxQuery  --static-172:27-->      github.com/influxdata/flux/parser.ParseSource`

Here is an (GNU) awk script that shows a count of call points where
1. Caller is not from flux
2. Call is static
3. Callee is from flux
4. Not the init function


On OSX you'll have to `brew install gawk` to use this.

```
{ 
    if ( match( $1, /\<influxdata\/flux\>/ ) == 0 && 
        match( $2, /\<static\>/ ) != 0 &&
        match( $3, /\<influxdata\/flux\>/ ) != 0 &&
        match( $3, /\<init\>$/ ) == 0 )
    { 
        calls[$3] += 1;
    } 
}
END {
    for ( key in calls ) {
        print calls[key], "\t", key;
}
```

[calls.txt](https://github.com/influxdata/influxdb/files/4503623/calls.txt)

**Comment by adrian-thurston** [comment]
Now have a program `calls` that can show us this.

https://github.com/influxdata/flux/tree/master/cmd/calls

**Comment by jsternberg** [comment]
I reran the above program on the `feat/use-algo-w` branch to encapsulate what we currently understand about influxdb since that will be the version we're merging in a week and basing the versioning work on. I am looking at the output with a focus on entrypoints which means I am filtering out some of the results if you were to run the call. For example, I am filtering out this call.

```
n38294:_/flux/arrow.NewBool  /Users/jsternberg/go/pkg/mod/github.com/influxdata/flux@v0.66.1-0.20200423173753-ffde07d2f718/arrow/bool.go:8:6
  n39335:(*_/influxdb/v2/storage/flux.booleanTable).toArrowBuffer  /Users/jsternberg/go/src/github.com/influxdata/influxdb/storage/flux/table.go:246:22
```

The reason is because this call is a result of an implementation detail. Some influxdb code uses some common code in flux to be able to implement a source, but this isn't actually a call as an entrypoint into flux even though it has all of the listed ways to identify a call into flux. I don't think we're going to be able to filter the difference between this call and a call to the parser using a static analysis program, so this manual filtering will probably be needed.

I put the original output and the filtered one [here](https://gist.github.com/jsternberg/eb3921166603fad9ce467955b1328bec/revisions).

The reason why I did this filtering is because this issue is not about versioning the go api but about versioning flux the language itself. This means I am filtering out calls that are used as part of implementing transformations.

There are a few basic categories that I have found.

1. Parsing a file or AST and then using or manipulating the AST. Parsing is considered a call into the library while manipulating the AST would be considered a Go API issue.
2. The `complete` package. As a caveat, the `complete` package is mostly dead code and we may want to decide to just drop it altogether. It doesn't seem to be costing us anything to keep it though.
3. Registering sources, transformations, planner rules, and builtin functions with the runtime. This seems to be the primary method that influxdb interacts with Flux outside of running Flux queries.
4. Evaluating an AST and executing the query in the Flux runtime. This is the most common entrypoint.

I do not think we have to write out a library isolating these calls. I think the information we have is enough to create a proposal for versioning that we can then prototype. The refactor that was done to isolate Flux dependencies seems to already have done most of this since the `FluxLanguageService`.

This is the `FluxLanguageService` as is currently in the algorithm w code.

```go
// FluxLanguageService is a service for interacting with flux code.
type FluxLanguageService interface {
	// Parse will take flux source code and produce a package.
	// If there are errors when parsing, the first error is returned.
	// An ast.Package may be returned when a parsing error occurs,
	// but it may be null if parsing didn't even occur.
	Parse(source string) (*ast.Package, error)

	// EvalAST will evaluate and run an AST.
	EvalAST(ctx context.Context, astPkg *ast.Package) ([]interpreter.SideEffect, values.Scope, error)

	// Completer will return a flux completer.
	Completer() complete.Completer
}
```
